crates/pera-core/src/signature_mpc/mpc_manager.rs

```diff
@@ -2,163 +2,231 @@ use crate::authority::authority_per_epoch_store::AuthorityPerEpochStore;
-        mut receiver: mpsc::Receiver<MPCInput>,
-    ) {
-        let consensus_adapter = Arc::clone(&self.consensus_adapter);
-        let epoch_store = self.epoch_store.clone();
-        let threshold = self.mpc_threshold_number_of_parties;
-        tokio::spawn(async move {
-            let mut party: ProofParty;
-            while let Some(message) = receiver.recv().await {
-                match message {
-                    MPCInput::InitEvent(_) => {
-                        party = match Self::handle_mpc_proof_init_event(
-                            public_parameters.clone(),
-                            consensus_adapter.clone(),
-                            epoch_store.clone(),
-                            threshold,
-                        )
-                        .await
-                        {
-                            Ok(party) => party,
-                            Err(err) => {
-                                // This should never happen, as there should be on-chain verification on the init transaction
-                                return;
-                            }
-                        };
-                    }
-                    MPCInput::Message => {
-                        // TODO (#235): Implement MPC messages handling
-                    }
-                }
+    fn advance(&mut self) -> Option<ConsensusTransaction> {
+        let party = mem::take(&mut self.party);
+        let Some(party) = party else {
+            // This should never happen, the party is initialized in the constructor
+            // and advance should not be called more than once simultaneously for the same instance
+            error!("Party is not initialized");
+            return None;
+        };
+        let Ok(advance_result) = party.advance(self.pending_messages.clone(), &(), &mut OsRng)
+        else {
+            // TODO (#263): Mark and punish the malicious validators that caused this advance to fail
+            self.pending_messages.clear();
+            return None;
+        };
+        match advance_result {
+            AdvanceResult::Advance((message, party)) => {
+                self.pending_messages.clear();
+                self.party = Some(party);
+                return self.new_signature_mpc_message(message);
+            }
+            AdvanceResult::Finalize(output) => {
+                // TODO (#238): Verify the output and write it to the chain
+                println!("Finalized output: {:?}", output);
+                self.status = MPCSessionStatus::Finished;
+            }
+        }
+        None
+    }
 
-                // TODO (#235): Implement MPC messages handling
+    fn new_signature_mpc_message(&self, message: ProofMessage) -> Option<ConsensusTransaction> {
+        let Some(epoch_store) = self.epoch_store.upgrade() else {
+            // TODO: (#259) Handle the case when the epoch switched in the middle of the MPC instance
+            return None;
+        };
```

Comment:
This code is repeated in other locations (e.g., lines 175–178), maybe during solving
[#259](https://github.com/dwallet-labs/dwallet-network/issues/259) this could be reformated as a single function so that
it is handled in the same way in different locations.

Reply:
yep, will do in #259